lib/ecto/repo/preloader.ex

```diff
@@ -252,13 +252,20 @@ defmodule Ecto.Repo.Preloader do
 
     # If we are returning many results, we must sort by the key too
     query =
-      case card do
-        :many ->
+      case {card, query.combinations} do
+        {:many, [{kind, _} | []]} ->
+          raise ArgumentError,
+                "#{kind} queries must be wrapped inside of a subquery " <>
+                  "when preloading a `has_many` or `many_to_many` association. " <>
+                  "You must also ensure that all members of the #{kind} query " <>
+                  "select the parent's foreign key"
```

Comment:
I was going to put an example in this error message because I could see someone being a bit confused. But I couldn't think of any example where I felt a combination preload was the best option and I felt weird advertising a query that I think shouldn't be used.